README.md
# enab-tech-blockchain
cybersecurity/blockchain presentation with impress.js 


docs/changelog.md
# CHANGELOG


## 14-09-2021
- gestione images-slide
- background con particles.js
- slide main menu
- slide ramo storia
- gestione stili con scss

## 12-05-2021
- firebase
- images and 3dmodels in public folder

## 12-05-2021
- first commit



public/index.js
(function() {
    
    var imp = impress();
    // console.log("impress", impress);
    window.imp = imp;
    
    imp.init();
    var root = document.querySelector("body");

    imp.lib.gc.addEventListener( root, "impress:stepenter", function( event, data ) {
        console.log("step enter");
        console.log("event", event);
        console.log("data", data);
    }, false );

    imp.lib.gc.addEventListener( root, "impress:stepleave", function( event ) {
        console.log("step leave");
    }, false );




    var loadModelOnEngine = function(params) {
        params = params || {};

        var domNode = params.domNode;
        var modelPath = params.modelPath;

        var engine3d = new Engine3d();

        engine3d.init(domNode);
    
        engine3d.start();
    
        engine3d.startRotate(-1);
    
        engine3d.loadGeometry(modelPath, function(err, node) {
            if (err) {
                return console.log(err);
            }
    
            node.getChildMeshes().forEach(function(m) {
                if(m.material) {
                    m.material.alpha = 0.6;
                }
            });

            // var material = engine3d.getEmissiveMaterialRGB({g: 150, b: 255, a: 1});
            // engine3d.setMeshMaterial(node, material);
    
    
        });


    };



    // loadModelOnEngine({
    //     domNode: document.querySelector("#menu canvas"),
    //     modelPath: "./3dmodels/lock.glb"
    // });


    // loadModelOnEngine({
    //     domNode: document.querySelector("#smart-city canvas"),
    //     modelPath: "./3dmodels/city1.glb"
    // });






    // Custom events

    // document.addEventListener( "app:test", function(event) {
        
    //     console.log("event", event);

    // }, false );


    // var event = document.createEvent( "CustomEvent" );
    // event.initCustomEvent( "app:test", true, true, {
    //     data: {id: 3}
    // });
    // window.dispatchEvent( event );


    





})();

tools/fsutils.js
const fs = require("fs");
const path = require("path");
const BBPromise = require("bluebird");



const readFile = BBPromise.promisify(fs.readFile);
const writeFile = BBPromise.promisify(fs.writeFile);
const appenFile = BBPromise.promisify(fs.appendFile);
const deleteFile = BBPromise.promisify(fs.unlink);
const rename = BBPromise.promisify(fs.rename);
const mkdir = BBPromise.promisify(fs.mkdir);
const stat = BBPromise.promisify(fs.stat);
const lstat = BBPromise.promisify(fs.lstat);
const readdir = BBPromise.promisify(fs.readdir);



const readJson = function(path) {
    return new BBPromise(function(resolve, reject) {
        fs.readFile(path, "utf8", function(err, text) {
            if (err) { return reject(err);}
            try {
                const result = JSON.parse(text);
                resolve(result);
            } catch (error) {
                reject(error);
            }
        });
    });
};


const readJsonSync = function(path) {
    try {
        const text = fs.readFileSync(path, "utf8");
        const result = JSON.parse(text);
        return result;
    } catch (error) {
        throw error;
    }
};


const writeJson = function(path, content) {
    return new BBPromise(function(resolve, reject) {
        const text = JSON.stringify(content, null, "    ");
        fs.writeFile(path, text, "utf8", function(err) {
            if (err) { return reject(err);}
            resolve();
        });
    });
};


const writeJsonSync = function(path, content) {
    const text = JSON.stringify(content, null, "    ");
    fs.writeFileSync(path, text);
};


const checkExist = function(path) {
    return new BBPromise(function (resolve) {
        fs.access(path, function (err) {
            if (err) { resolve(false); } 
            else { resolve(true); }
        });
    });
};


const checkExistSync = function(path) {
    try {
        fs.accessSync(path);
        return true;
    } catch (error) {
        return false;
    }
};


const ensureDir = (dirpath) => {
    return new BBPromise((resolve, reject) => {
        fs.mkdir(dirpath, err => {
            if (err && err.code != 'EEXIST') { }
            resolve();
        });
    });
};


const ensureDirSync = function(dirpath) {
    if (!fs.existsSync(dirpath)){
        fs.mkdirSync(dirpath);
    }
};


const copyFile = function (src, dest) {
    return new BPromise(function (resolve, reject) {
        const readStream = fs.createReadStream(src);
        const writeStream = fs.createWriteStream(dest);
        readStream.on('error', function (err) {
            reject(err);
        });
        writeStream.on('error', function (err) {
            reject(err);
        });
        writeStream.on('close', function () {
            resolve();
        });
        readStream.pipe(writeStream);
    });
};



// ##TODO capire se hanno senso metodi semplificati o meno
const isFile = function (filePath) {
    return new Promise(function (resolve, reject) {
        fs.lstat(filePath, function (err, stats) {
            if (err) { return reject(err); }
            resolve(stats.isFile());
        });
    });
};


const getDirFileList = async function(dirPath) {
    try {
        
        const elements = await readdir(dirPath);
        const files = [];

        for (let i = 0; i < elements.length; i++) {
            const element = elements[i];

            const elementPath = path.join(dirPath, element);

            const stats = await lstat(elementPath);
            
            const isFile = stats.isFile();

            if (isFile) { files.push(element); }

        }

        return files;

    } catch (error) {
        throw error;
    }
};


const getDirFileListSync = function(dirPath) {
    try {
        
        const elements = fs.readdirSync(dirPath);
        const files = [];

        for (let i = 0; i < elements.length; i++) {
            const element = elements[i];

            const elementPath = path.join(dirPath, element);

            const stats = fs.lstatSync(elementPath);
            
            const isFile = stats.isFile();

            if (isFile) { files.push(element); }

        }

        return files;

    } catch (error) {
        throw error;
    }
};









module.exports = {
    // fs promisificati
    readFile,
    writeFile,
    appenFile,
    deleteFile,
    rename,
    mkdir,
    stat,
    lstat,
    readdir,

    // new helpers
    readJson,
    readJsonSync,
    writeJson,
    writeJsonSync,
    checkExist,
    checkExistSync,
    ensureDir,
    ensureDirSync,
    copyFile,

    getDirFileList,
    getDirFileListSync

};

public/js/bg__T1.js


(function() {

    
    ////////////////////////// PARTICLE ENGINE ////////////////////////////////////////
    /////////////////////////////////////////////////////////////////////////////////////////

    var ParticleEngine = (function() {
        'use strict';

        function ParticleEngine(canvas_id) {
            // enforces new
            if (!(this instanceof ParticleEngine)) {
                return new ParticleEngine(args);
            }
            
            var _ParticleEngine = this;

            this.canvas_id = canvas_id;
            this.stage = new createjs.Stage(canvas_id);
            this.totalWidth = this.canvasWidth = document.getElementById(canvas_id).width = document.getElementById(canvas_id).offsetWidth;
            this.totalHeight = this.canvasHeight = document.getElementById(canvas_id).height = document.getElementById(canvas_id).offsetHeight;
            this.compositeStyle = "lighter";

            // this.particleSettings = [{id:"small", num:300, fromX:0, toX:this.totalWidth, ballwidth:3, alphamax:0.4, areaHeight:.5, color:"#0cdbf3", fill:false}, 
            //                         {id:"medium", num:100, fromX:0, toX:this.totalWidth,  ballwidth:8, alphamax:0.3, areaHeight:1, color:"#6fd2f3", fill:true}, 
            //                         {id:"large", num:10, fromX:0, toX:this.totalWidth, ballwidth:30,  alphamax:0.2, areaHeight:1, color:"#93e9f3", fill:true}];

            this.particleSettings = [
                {id:"small", num:300, fromX:0, toX:this.totalWidth, ballwidth:3, alphamax:0.4, areaHeight:0.5, color:"#444", fill:false}, 
                {id:"medium", num:100, fromX:0, toX:this.totalWidth,  ballwidth:8, alphamax:0.3, areaHeight:1, color:"#444", fill:true}, 
                {id:"large", num:10, fromX:0, toX:this.totalWidth, ballwidth:30,  alphamax:0.2, areaHeight:1, color:"#444", fill:true}
            ];

            this.particleArray = [];
            // this.lights = [{ellipseWidth:400, ellipseHeight:100, alpha:0.6, offsetX:0, offsetY:0, color:"#6ac6e8"}, 
            //                 {ellipseWidth:350, ellipseHeight:250, alpha:0.3, offsetX:-50, offsetY:0, color:"#54d5e8"}, 
            //                 {ellipseWidth:100, ellipseHeight:80, alpha:0.2, offsetX:80, offsetY:-50, color:"#2ae8d8"}];
            this.lights = [
                {ellipseWidth:400, ellipseHeight:100, alpha:0.6, offsetX:0, offsetY:0, color:"#eee"}, 
                {ellipseWidth:350, ellipseHeight:250, alpha:0.3, offsetX:-50, offsetY:0, color:"#eee"}, 
                {ellipseWidth:100, ellipseHeight:80, alpha:0.2, offsetX:80, offsetY:-50, color:"#eee"}
            ];

            this.stage.compositeOperation = _ParticleEngine.compositeStyle;


            function drawBgLight()
            {
                var light;
                var bounds;
                var blurFilter;
                for (var i = 0, len = _ParticleEngine.lights.length; i < len; i++) {				
                    light = new createjs.Shape();
                    light.graphics.beginFill(_ParticleEngine.lights[i].color).drawEllipse(0, 0, _ParticleEngine.lights[i].ellipseWidth, _ParticleEngine.lights[i].ellipseHeight);
                    light.regX = _ParticleEngine.lights[i].ellipseWidth/2;
                    light.regY = _ParticleEngine.lights[i].ellipseHeight/2; 
                    light.y = light.initY = _ParticleEngine.totalHeight/2 + _ParticleEngine.lights[i].offsetY;
                    light.x = light.initX =_ParticleEngine.totalWidth/2 + _ParticleEngine.lights[i].offsetX;

                    blurFilter = new createjs.BlurFilter(_ParticleEngine.lights[i].ellipseWidth, _ParticleEngine.lights[i].ellipseHeight, 1);
                    bounds = blurFilter.getBounds();
                    light.filters = [blurFilter];
                    light.cache(bounds.x-_ParticleEngine.lights[i].ellipseWidth/2, bounds.y-_ParticleEngine.lights[i].ellipseHeight/2, bounds.width*2, bounds.height*2);
                    light.alpha = _ParticleEngine.lights[i].alpha;

                    light.compositeOperation = "screen";
                    _ParticleEngine.stage.addChildAt(light, 0);

                    _ParticleEngine.lights[i].elem = light;
                }

                TweenMax.fromTo(_ParticleEngine.lights[0].elem, 10, {scaleX:1.5, x:_ParticleEngine.lights[0].elem.initX, y:_ParticleEngine.lights[0].elem.initY},{yoyo:true, repeat:-1, ease:Power1.easeInOut, scaleX:2, scaleY:0.7});
                TweenMax.fromTo(_ParticleEngine.lights[1].elem, 12, { x:_ParticleEngine.lights[1].elem.initX, y:_ParticleEngine.lights[1].elem.initY},{delay:5, yoyo:true, repeat:-1, ease:Power1.easeInOut, scaleY:2, scaleX:2, y:_ParticleEngine.totalHeight/2-50, x:_ParticleEngine.totalWidth/2+100});
                TweenMax.fromTo(_ParticleEngine.lights[2].elem, 8, { x:_ParticleEngine.lights[2].elem.initX, y:_ParticleEngine.lights[2].elem.initY},{delay:2, yoyo:true, repeat:-1, ease:Power1.easeInOut, scaleY:1.5, scaleX:1.5, y:_ParticleEngine.totalHeight/2, x:_ParticleEngine.totalWidth/2-200});
            }
            
            var blurFilter;
            function drawParticles(){

                for (var i = 0, len = _ParticleEngine.particleSettings.length; i < len; i++) {
                    var ball = _ParticleEngine.particleSettings[i];

                    var circle;
                    for (var s = 0; s < ball.num; s++ )
                    {
                        circle = new createjs.Shape();
                        if(ball.fill){
                            circle.graphics.beginFill(ball.color).drawCircle(0, 0, ball.ballwidth);
                            blurFilter = new createjs.BlurFilter(ball.ballwidth/2, ball.ballwidth/2, 1);
                            circle.filters = [blurFilter];
                            var bounds = blurFilter.getBounds();
                            circle.cache(-50+bounds.x, -50+bounds.y, 100+bounds.width, 100+bounds.height);
                        }else{
                            circle.graphics.beginStroke(ball.color).setStrokeStyle(1).drawCircle(0, 0, ball.ballwidth);
                        }
                        
                        circle.alpha = range(0, 0.1);
                        circle.alphaMax = ball.alphamax;
                        circle.distance = ball.ballwidth * 2;
                        circle.ballwidth = ball.ballwidth;
                        circle.flag = ball.id;
                        _ParticleEngine.applySettings(circle, ball.fromX, ball.toX, ball.areaHeight);
                        circle.speed = range(2, 10);
                        circle.y = circle.initY;
                        circle.x = circle.initX;
                        circle.scaleX = circle.scaleY = range(0.3, 1);

                        _ParticleEngine.stage.addChild(circle);
                        

                        animateBall(circle);

                        _ParticleEngine.particleArray.push(circle);
                    }
                }	
            }

            this.applySettings = function(circle, positionX, totalWidth, areaHeight)
            {
                circle.speed = range(1, 3);
                circle.initY = weightedRange(0, _ParticleEngine.totalHeight , 1, [_ParticleEngine.totalHeight * (2-areaHeight/2)/4, _ParticleEngine.totalHeight*(2+areaHeight/2)/4], 0.8 );
                circle.initX = weightedRange(positionX, totalWidth, 1, [positionX+ ((totalWidth-positionX))/4, positionX+ ((totalWidth-positionX)) * 3/4], 0.6);
            }

            function animateBall(ball)
            {
                var scale = range(0.3, 1);
                var xpos = range(ball.initX - ball.distance, ball.initX + ball.distance);
                var ypos = range(ball.initY - ball.distance, ball.initY + ball.distance);
                var speed = ball.speed;
                TweenMax.to(ball, speed, {scaleX:scale, scaleY:scale, x:xpos, y:ypos, onComplete:animateBall, onCompleteParams:[ball], ease:Cubic.easeInOut});	
                TweenMax.to(ball, speed/2, {alpha:range(0.1, ball.alphaMax), onComplete:fadeout, onCompleteParams:[ball, speed]});	
            }	

            function fadeout(ball, speed)
            {
                ball.speed = range(2, 10);
                TweenMax.to(ball, speed/2, {alpha:0 });
            }

            drawBgLight();
            drawParticles();
        }

        ParticleEngine.prototype.render = function()
        {
            this.stage.update();
        }

        ParticleEngine.prototype.resize = function()
        {
            this.totalWidth = this.canvasWidth = document.getElementById(this.canvas_id).width = document.getElementById(this.canvas_id).offsetWidth;
            this.totalHeight = this.canvasHeight = document.getElementById(this.canvas_id).height = document.getElementById(this.canvas_id).offsetHeight;
            this.render();

            for (var i= 0, length = this.particleArray.length; i < length; i++)
            {
                this.applySettings(this.particleArray[i], 0, this.totalWidth, this.particleArray[i].areaHeight);
            }

            for (var j = 0, len = this.lights.length; j < len; j++) {
                this.lights[j].elem.initY = this.totalHeight/2 + this.lights[j].offsetY;
                this.lights[j].elem.initX =this.totalWidth/2 + this.lights[j].offsetX;
                TweenMax.to(this.lights[j].elem, .5, {x:this.lights[j].elem.initX, y:this.lights[j].elem.initY});			
            }
        }

        return ParticleEngine;

    }());


    ////////////////////////UTILS//////////////////////////////////////
    //////////////////////////////////////////////////////////////////

    function range(min, max)
    {
        return min + (max - min) * Math.random();
    }
            
    function round(num, precision)
    {
    var decimal = Math.pow(10, precision);
    return Math.round(decimal* num) / decimal;
    }

    function weightedRange(to, from, decimalPlaces, weightedRange, weightStrength)
    {
        if (typeof from === "undefined" || from === null) { 
            from = 0; 
        }
        if (typeof decimalPlaces === "undefined" || decimalPlaces === null) { 
            decimalPlaces = 0; 
        }
        if (typeof weightedRange === "undefined" || weightedRange === null) { 
            weightedRange = 0; 
        }
        if (typeof weightStrength === "undefined" || weightStrength === null) { 
            weightStrength = 0; 
        }

    var ret
    if(to == from){return(to);}
    
    if(weightedRange && Math.random()<=weightStrength){
        ret = round( Math.random()*(weightedRange[1]-weightedRange[0]) + weightedRange[0], decimalPlaces )
    }else{
        ret = round( Math.random()*(to-from)+from, decimalPlaces )
    }
    return(ret);
    }

    ///////////////// RUN CODE //////////////////////////
    //////////////////////////////////////////////////////

    var particles
    (function(){
        particles = new ParticleEngine('projector');
        createjs.Ticker.addEventListener("tick", updateCanvas);
        window.addEventListener('resize', resizeCanvas, false);

        function updateCanvas(){
            particles.render();
        }

        function resizeCanvas(){
            particles.resize();
        }
    }());





})();



public/js/engine3d.js
var Engine3d = (function() {
    

    var ifUndefDefault = function (value, defaultValue) {
        return value !== undefined ? value : defaultValue;
    };


    var Engine3d = (function() {
    
        function Engine3d() {
            this.canvas = null;
            this.engine = null;
            this.scene = null;
        }

        Engine3d.prototype.createScene = function(engine, canvas) {
            
            var radius = 1;
            var alpha =  Math.PI/2;
            var beta = Math.PI/2;
            
            var scene = new BABYLON.Scene(engine);
            
            // var camera = new BABYLON.FreeCamera('camera1', new BABYLON.Vector3(0, 5, -10), scene);
            var camera = new BABYLON.ArcRotateCamera("Camera", alpha, beta, radius, BABYLON.Vector3.Zero(), scene);
            camera.attachControl(canvas, true);
            camera.minZ = 0.1;
            camera.lowerRadiusLimit = 0.5;
            camera.upperRadiusLimit = 1;
    
            camera.wheelPrecision = 100;
    
            camera.setTarget(BABYLON.Vector3.Zero());
            
            // camera.attachControl(canvas, false, false);
            camera.attachControl(canvas, false, true, 2);
    
    
            var light = new BABYLON.HemisphericLight('light1', new BABYLON.Vector3(0, 1, 0), scene);
            
            return scene;
        };

        Engine3d.prototype.init = function(canvas, params) {
            var _this = this;

            this.canvas = canvas;
            
            var engine = new BABYLON.Engine(this.canvas, true, {
                preserveDrawingBuffer: true, 
                stencil: true
            });
            this.engine = engine;
            
            var scene = this.createScene(engine, this.canvas);
            this.scene = scene;
    
            // this.showWorldAxis(scene, 4);
    
            scene.clearColor = new BABYLON.Color4(0, 0, 0, 0);
                        
            /* Resize event */
            window.addEventListener('resize', function(){
                _this.resize();
            });

            this.canvas.addEventListener("wheel", function(evt) {
                evt.preventDefault();
            });

        };


        Engine3d.prototype.start = function() {
            var _this = this;
            this.engine.runRenderLoop(function(){
                _this.scene.render();
            });
            this.engine.resize();
        };


        Engine3d.prototype.stop = function() {
            this.engine.stopRenderLoop();
        };


        Engine3d.prototype.resize = function() {
            this.engine.resize();
        };


        var _decodePath = function (path) {
            var fileStart = path.lastIndexOf('/') + 1;
            var fileName = path.substring(fileStart);
            var filePath = path.substring(0, fileStart);
            return [filePath, fileName];
        };
        var _importMesh = function(path, scene, onSuccess, onProgress, onError, pluginExtension) {
            var decoded = _decodePath(path);
            var dirPath = decoded[0];
            var fileName = decoded[1];

            BABYLON.SceneLoader.ImportMesh(
                null, // mesh to import, null for all
                dirPath,
                fileName,
                scene,
                onSuccess,
                onProgress,
                onError,
                pluginExtension
            );
            
        };
        Engine3d.prototype.loadGeometry = function(path, callback) {

            var node = new BABYLON.TransformNode("", this.scene);

            _importMesh(path, this.scene, function(meshes) {
                console.log("meshes", meshes);

                meshes.forEach(function(m) { m.parent = node; });
                callback(null, node);
            },
            null,
            function(err) {
                callback(err);
            });


            // node.rotate(BABYLON.Axis.Z, BABYLON.Tools.ToRadians(90), BABYLON.Space.LOCAL);



        };


        Engine3d.prototype.startRotate = function(node) {
            var scene = this.scene;
            var rotationHandler = function() {
                scene.activeCamera.alpha += BABYLON.Tools.ToRadians(-0.5);
                // node.rotate(BABYLON.Axis.X, BABYLON.Tools.ToRadians(-0.5), BABYLON.Space.LOCAL);
            };
            this.rotationHandler = rotationHandler;
            this.scene.registerBeforeRender(rotationHandler);
        };
        Engine3d.prototype.stopRotate = function() {
            this.scene.unregisterBeforeRender(rotationHandler);
            this.rotationHandler = null;
        };


        Engine3d.prototype.getEmissiveMaterialRGB = function (color) {
            color = color || {};
            var scene = this.scene;

            var r = ifUndefDefault(color.r, 0);
            var g = ifUndefDefault(color.g, 0);
            var b = ifUndefDefault(color.b, 0);
            var a = ifUndefDefault(color.a, 1);

            var materialName = "mat_emis_rgba_" + r + "_" + g + "_" + b + "_" + a;
            
            var material = scene.getMaterialByName(materialName);
    
            if (material) { return material; }
            
            material = new BABYLON.StandardMaterial(materialName, scene);
            material.emissiveColor = new BABYLON.Color3(r / 255, g / 255, b / 255);
            material.disableLighting = true;
            material.alpha = a;
            
            return material;
        };


        Engine3d.prototype.setMeshMaterial = function(mesh, material) {
            var meshes = mesh.getChildMeshes();
            meshes.push(mesh);
            meshes.forEach(function(m) {
                if (m.geometry) {
                    m.material = material;
                }
            });
        };


        return Engine3d;

    })();

    
    return Engine3d;

})();

public/js/bg.js
(function () {

    particlesJS('bg_animated',
        {
            "particles": {
                "number": {
                    "value": 80,
                    "density": {
                        "enable": true,
                        "value_area": 2000
                    }
                },
                "color": {
                    "value": "#aaa"
                },
                "shape": {
                    "type": "circle",
                    "stroke": {
                        "width": 0,
                        "color": "#444"
                    },
                    "polygon": {
                        "nb_sides": 5
                    },
                    "image": {
                        "src": "img/github.svg",
                        "width": 100,
                        "height": 100
                    }
                },
                "opacity": {
                    "value": 0.5,
                    "random": false,
                    "anim": {
                        "enable": false,
                        "speed": 1,
                        "opacity_min": 0.1,
                        "sync": false
                    }
                },
                "size": {
                    "value": 5,
                    "random": true,
                    "anim": {
                        "enable": false,
                        "speed": 40,
                        "size_min": 0.1,
                        "sync": false
                    }
                },
                "line_linked": {
                    "enable": true,
                    "distance": 150,
                    "color": "#444",
                    "opacity": 0.4,
                    "width": 1
                },
                "move": {
                    "enable": true,
                    "speed": 6,
                    "direction": "none",
                    "random": false,
                    "straight": false,
                    "out_mode": "out",
                    "attract": {
                        "enable": false,
                        "rotateX": 600,
                        "rotateY": 1200
                    }
                }
            },
            "interactivity": {
                "detect_on": "canvas",
                "events": {
                    "onhover": {
                        "enable": false,
                        "mode": "repulse"
                    },
                    "onclick": {
                        "enable": true,
                        "mode": "push"
                    },
                    "resize": true
                },
                "modes": {
                    "grab": {
                        "distance": 400,
                        "line_linked": {
                            "opacity": 1
                        }
                    },
                    "bubble": {
                        "distance": 400,
                        "size": 40,
                        "duration": 2,
                        "opacity": 8,
                        "speed": 3
                    },
                    "repulse": {
                        "distance": 200
                    },
                    "push": {
                        "particles_nb": 4
                    },
                    "remove": {
                        "particles_nb": 2
                    }
                }
            },
            "retina_detect": true,
            "config_demo": {
                "hide_card": false,
                "background_color": "#b61924",
                "background_image": "",
                "background_position": "50% 50%",
                "background_repeat": "no-repeat",
                "background_size": "cover"
            }
        }

    );

})();